rust-runtime/aws-smithy-compression/src/http.rs

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Checksum support for HTTP requests and responses.
+
+/// Support for the `http-body-0-4` and `http-0-2` crates.
+#[cfg(feature = "http-body-0-4-x")]
+pub mod http_body_0_4_x {
+    use crate::Compression;
+    use http_0_2::header::{HeaderMap, HeaderName, HeaderValue};
+
+    /// Implementors of this trait can be used to compress HTTP requests.
+    pub trait RequestCompressor: Compression + Send + Sync {
+        /// Return a map of headers that must be included when a request is compressed.
+        fn headers(self: Box<Self>) -> HeaderMap<HeaderValue> {
```

Comment:
Can you help better understand the API signature, i.e. the use of `Box<Self>` and `HeaderMap<HeaderValue>` as opposed to say `fn headers(&self) -> (&HeaderName, &HeaderValue)`?

Reply:
I refactored this code and related stuff to work more simply.

Headers now get appended like this:
```rust
fn wrap_request_body_in_compressed_body(
    request: &mut HttpRequest,
    request_compress_impl: Box<dyn CompressRequest>,
) -> Result<(), BuildError> {
    request.headers_mut().append(
        request_compress_impl.header_name(),
        request_compress_impl.header_value(),
    );
    let mut body = {
        let body = mem::replace(request.body_mut(), SdkBody::taken());
        body.map(move |body| {
            let body = CompressedBody::new(
                body,
                request_compress_impl.clone(),
            );

            SdkBody::from_body_0_4(body)
        })
    };
    mem::swap(request.body_mut(), &mut body);

    Ok(())
}
```